core/backend/src/IModelDb.ts

```diff
@@ -1082,6 +1094,148 @@ export abstract class IModelDb extends IModel {
+      // This is always a fatal error - it can happen only if the app started with a briefcase that is behind the tip and then uses the same primary key values (e.g., ElementIds)
+      // that have already been used by some other app using the SAME briefcase ID that recently pushed changes. That can happen only if the app makes changes without first pulling and acquiring locks.
+      if (!args.hasLocalChanges) {
+        // This changeset is bad. However, it is already in the timeline. We must allow services such as
+        // checkpoint-creation, change history, and other apps to apply any changeset that is in the timeline.
+        Logger.logWarning(category, "PRIMARY KEY INSERT CONFLICT - resolved by replacing the existing row with the incoming row");
+        args.dump();
+      } else {
+        const msg = "PRIMARY KEY INSERT CONFLICT - rejecting this changeset";
+        args.setLastError(msg);
+        Logger.logError(category, msg);
+        args.dump();
+        return DbConflictResolution.Abort;
+      }
+    }
+
+    if (args.cause === DbConflictCause.ForeignKey) {
+      // Note: No current or conflicting row information is provided if it's a FKey conflict
+      // Since we abort on FKey conflicts, always try and provide details about the error
+      const nConflicts = args.getForeignKeyConflicts();
+
+      // Note: There is no performance implication of follow code as it happen toward end of
+      // apply_changeset only once so we be querying value for 'DebugAllowFkViolations' only once.
+      if (this.nativeDb.queryLocalValue("DebugAllowFkViolations")) {
+        Logger.logError(category, `Detected ${nConflicts} foreign key conflicts in changeset. Continuing merge as 'DebugAllowFkViolations' flag is set. Run 'PRAGMA foreign_key_check' to get list of violations.`);
+        return DbConflictResolution.Skip;
+      } else {
+        const msg = `Detected ${nConflicts} foreign key conflicts in ChangeSet. Aborting merge.`;
+        args.setLastError(msg);
+        return DbConflictResolution.Abort;
+      }
+    }
+
+    if (args.cause === DbConflictCause.NotFound) {
+      /*
+       * Note: If DbConflictCause = NotFound, the primary key was not found, and returning DbConflictResolution::Replace is
+       * not an option at all - this will cause a BE_SQLITE_MISUSE error.
+       */
+      if (args.opcode === DbOpcode.Delete) {
+        // Caused by CASCADE DELETE on a foreign key, and is usually not a problem.
+        return DbConflictResolution.Skip;
+      }
+
+      if (args.opcode === DbOpcode.Update && args.tableName.startsWith("ec_")) {
+        // Caused by a ON DELETE SET NULL constraint on a foreign key - this is known to happen with "ec_" tables, but needs investigation if it happens otherwise
+        return DbConflictResolution.Skip;
+      }
+
+      // Refer to comment below
```

Comment:
Can you explain this more?

If the specific change in this changeset is an Update to some row, then we skip the incoming change? Does that correspond to the point about 'briefcase with local changes' in the farther down comment?

Not sure I understand

Reply:
No, it means there is a conflict where the parent row is missing. e.g. we update a row that has fk and fk point to a row that does not exist. then we have to decide what to do. It could be an insert/update/delete operation that can use NotFound conflict.

Generally, we will skip on not found since inserting it will later cause an FK violation as we attempt to insert/update an orphan row. Though replacing for insert will cause an FK violation later.

Reply:
So even though one of your comments says "// Caused by a ON DELETE SET NULL constraint on a foreign key - this is known to happen with "ec_" tables, but needs investigation if it happens otherwise"

You still return the same DbConflictResolution.Skip `if args.tableName.startsWith("ec_")` evaluates to false. How will this investigation take place?

Within the same larger if statement `(if args.cause === DbConflictcause.NotFound) ` There is also this comment:

"/*
       * Note: If DbConflictCause = NotFound, the primary key was not found, and returning DbConflictResolution::Replace is
       * not an option at all - this will cause a BE_SQLITE_MISUSE error.
       */"
       
There is also this comment which says returning DbConflictResolution::Replace is not an option at all, but then you return replace in `
      return args.opcode === DbOpcode.Update ? DbConflictResolution.Skip : DbConflictResolution.Replace;`
Does that comment only apply to the opcode of delete?

Reply:
My mistake, I have to say that a better source is the following. For NotFound replace is not an option. That will make it insert so sqlite will return Misuse which is correct. But NotFound is only ever going to happen for Update/Delete. So i guess though check is there it will never happen.

I copied the conflict resolver from the native side as is. I didn't want to make any changes to it. 

https://www.sqlite.org/session/sqlite3changeset_apply.html

Reply:
So after adding test i have update the conflict handler on `NotFound` the only thing we can do is `Skip`